roachprod: changes in promhelper client for modify labels

nameisbhaskar opened the pull request:
When a test is run, a new label has to be added to prometheus. To do this, we need to first get the current config, add the new label and update the config. This change does that. Post this change, the label update will be integrated.

The change also needs a change in httputil client GET function to support headers being passed.

Fixes: #124319
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124460)
<!-- Reviewable:end -->

nameisbhaskar committed:
roachprod: changes in promhelper client for modify labels

When a test is run, a new label has to be added to prometheus.
To do this, we need to first get the current config, add the
new label and update the config. This change does that. Post
this change, the label update will be integrated.

The change also needs a change in httputil client GET function
to support headers being passed.

Fixes: #124319
Epic: none

DarrylWong commented on pkg/util/httputil/client.go:
```diff
@@ -63,8 +63,8 @@ type Client struct {
 // It also uses the default client with a default 3 second timeout.
-func Get(ctx context.Context, url string) (resp *http.Response, err error) {
-	return DefaultClient.Get(ctx, url)
+func Get(ctx context.Context, url string, h *http.Header) (resp *http.Response, err error) {
```
Assuming that you're adding `GetWithHeaders` to avoid breaking old code that uses `Get`, I don't think we should change the function signature here. It should just call `DefaultClient.Get` and you can add a `GetWithHeaders` function separately if needed. That should get rid of the diffs in non roachprod/roachtest stuff.

nameisbhaskar replied:
I initially thought about this, but given that the usage was less, I modified the method signature rather than maintaining 2 methods. Also, others like Put does not have 2 methods.

DarrylWong commented on pkg/roachprod/roachprod.go:
```diff
@@ -1565,6 +1577,20 @@ func RemoveLabels(l *logger.Logger, clusterName string, labels []string) error {
 		}
 	}
+
+	nodeLabels := make(map[int][]string)
```
`RemoveLabels` is for removing VM labels, should this be in it's own function?

DarrylWong commented on pkg/roachprod/roachprod.go:
```diff
@@ -1542,6 +1542,18 @@ func AddLabels(l *logger.Logger, clusterName string, labels map[string]string) e
 			m.Labels[k] = v
 		}
 	}
+	nodeLabels := make(map[int]map[string]string)
```
Same as comment about `RemoveLabels`.

DarrylWong reviewed (commented):
Can you explain to me why we need to update the config this way? By the time we call `cluster.Start` the first time, we should already the test name and the run ID, is there a reason we can't add it all in one shot?

nameisbhaskar commented:
> Can you explain to me why we need to update the config this way? By the time we call `cluster.Start` the first time, we should already the test name and the run ID, is there a reason we can't add it all in one shot?

If a cluster is reused, do we always stop the nodes and then add the tags and start? If that is the case, we should be good without this change.

DarrylWong commented:
Yes, that's correct.

If we reuse a cluster, we call `c.WipeForReuse`.

https://github.com/cockroachdb/cockroach/blob/99d7687ba82a614a509c62a12bed411de006d05b/pkg/cmd/roachtest/test_runner.go#L636-L637

This eventually calls `c.Stop`

https://github.com/cockroachdb/cockroach/blob/6918532c343556a4dab10de0c9162f73fa712069/pkg/roachprod/install/cluster_synced.go#L547-L549

Same with removing labels. We call this at the end of `runTest`

https://github.com/cockroachdb/cockroach/blob/99d7687ba82a614a509c62a12bed411de006d05b/pkg/cmd/roachtest/test_runner.go#L1031-L1033

nameisbhaskar commented:
Thats great! Then we should be good without this change. I will close this PR! Thanks for pointing out!!
